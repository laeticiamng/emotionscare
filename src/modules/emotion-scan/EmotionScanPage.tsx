"use client";
import React from "react";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { PageHeader, Card, Button, ProgressBar, Sparkline } from "@/COMPONENTS.reg"; // ProgressBar/Sparkline ajoutés en P6
import { recordEvent } from "@/lib/scores/events"; // si P6 non intégré, cet import peut être ignoré (no-op)
import { logger } from "@/lib/logger";
import { useAuth } from "@/contexts/AuthContext";
import { useError } from "@/contexts";
import { LoadingState } from "@/components/loading/LoadingState";
import { UnifiedEmptyState } from "@/components/ui/unified-empty-state";
import { History } from "lucide-react";
import {
  invokeEmotionScan,
  getEmotionScanHistory,
  summarizeLikertAnswers,
  deriveScore10,
  EmotionScanHistoryEntry,
  EmotionAnalysisResult,
  persistEmotionScanResult,
} from "@/services/emotionScan.service";

const POS = [
  { id: "active", label: "Actif(ve)" },
  { id: "determined", label: "Déterminé(e)" },
  { id: "attentive", label: "Attentif(ve)" },
  { id: "inspired", label: "Inspiré(e)" },
  { id: "alert", label: "Alerte" }
] as const;

const NEG = [
  { id: "upset", label: "Contrarié(e)" },
  { id: "hostile", label: "Hostile" },
  { id: "ashamed", label: "Honteux(se)" },
  { id: "nervous", label: "Nerveux(se)" },
  { id: "afraid", label: "Effrayé(e)" }
] as const;

type Likert = 1|2|3|4|5;
type Resp = Record<(typeof POS[number] | typeof NEG[number])["id"], Likert | undefined>;

type EmotionScanMutationVariables = {
  text: string;
  context?: string;
  previousEmotions?: Record<string, number>;
  accessToken?: string | null;
  pa: number;
  na: number;
  balance: number;
};

const LOCAL_HISTORY_KEY = "emotion_scan_history_v2";

function loadLocalHistory(): number[] {
  try {
    const payload = localStorage.getItem(LOCAL_HISTORY_KEY);
    if (!payload) return [];
    const parsed = JSON.parse(payload);
    return Array.isArray(parsed) ? parsed.filter((value: unknown) => typeof value === "number") : [];
  } catch {
    return [];
  }
}

function saveLocalHistory(values: number[]) {
  try {
    localStorage.setItem(LOCAL_HISTORY_KEY, JSON.stringify(values.slice(-12)));
  } catch {
    // ignore storage errors (private mode, quota, etc.)
  }
}

export default function EmotionScanPage() {
  const [resp, setResp] = React.useState<Resp>({} as any);
  const [submitted, setSubmitted] = React.useState(false);
  const [analysis, setAnalysis] = React.useState<EmotionAnalysisResult | null>(null);
  const [errorMessage, setErrorMessage] = React.useState<string | null>(null);
  const [localHistory, setLocalHistory] = React.useState<number[]>(() => loadLocalHistory());

  const { user, session } = useAuth();
  const queryClient = useQueryClient();
  const { addError } = useError();

  React.useEffect(() => {
    if (!user?.id) {
      setLocalHistory(loadLocalHistory());
    }
  }, [user?.id]);

  const { data: remoteHistory = [], isLoading: historyLoading } = useQuery({
    queryKey: ["emotion-scan-history", user?.id],
    queryFn: () => getEmotionScanHistory(user!.id, 12),
    enabled: Boolean(user?.id),
    staleTime: 60 * 1000,
    onError: (error: unknown) => {
      const message = error instanceof Error ? error.message : "Impossible de charger l'historique des scans";
      addError({
        message,
        severity: "medium",
        stack: error instanceof Error ? error.stack : undefined,
        context: {
          scope: "emotion-scan-history",
          userId: user?.id ?? "anonymous",
        },
      });
    },
  });

  const historyEntries: EmotionScanHistoryEntry[] = React.useMemo(
    () => (user?.id ? remoteHistory : []),
    [remoteHistory, user?.id]
  );

  const sparklineValues = React.useMemo(() => {
    if (user?.id) {
      return [...historyEntries].reverse().map(entry => entry.normalizedBalance);
    }
    return localHistory;
  }, [historyEntries, localHistory, user?.id]);

  const latestScores = historyEntries[0]?.scores;

  const emotionScanMutation = useMutation({
    mutationFn: (variables: EmotionScanMutationVariables) =>
      invokeEmotionScan(
        {
          text: variables.text,
          context: variables.context,
          previousEmotions: variables.previousEmotions,
        },
        { accessToken: variables.accessToken },
      ),
    onMutate: () => {
      setErrorMessage(null);
    },
    onSuccess: async (result, variables) => {
      setAnalysis(result);

      if (user?.id) {
        if (!result.persisted) {
          const summary = [
            `Émotion dominante: ${result.dominantEmotion}`,
            `Confiance: ${result.confidence}%`,
            `Équilibre émotionnel: ${result.emotionalBalance}/100`,
            `Balance I-PANAS-SF: ${variables.balance >= 0 ? "+" : ""}${variables.balance}`,
          ].join(" · ");

          try {
            await persistEmotionScanResult({
              userId: user.id,
              scanType: "self-report",
              mood: result.dominantEmotion,
              confidence: result.confidence,
              summary,
              recommendations: result.recommendations,
              insights: result.insights,
              emotions: result.emotions,
              emotionalBalance: result.emotionalBalance,
              context: variables.context,
              previousEmotions: variables.previousEmotions ?? null,
            });
          } catch (persistError) {
            logger.error("Failed to persist emotion scan result", persistError, "emotion-scan.ui");
            const message =
              persistError instanceof Error
                ? persistError.message
                : "Impossible d'enregistrer le résultat du scan émotionnel";
            addError({
              message,
              severity: "medium",
              stack: persistError instanceof Error ? persistError.stack : undefined,
              context: {
                scope: "emotion-scan-persist",
                userId: user?.id ?? "anonymous",
              },
            });
          }
        }

        await Promise.all([
          queryClient.invalidateQueries({ queryKey: ["emotion-scan-history", user.id] }),
          queryClient.invalidateQueries({ queryKey: ["recent-scans", user.id] }),
        ]);
      } else {
        setLocalHistory(prev => {
          const next = [...prev, result.emotionalBalance].slice(-12);
          saveLocalHistory(next);
          return next;
        });
      }
    },
    onError: (error: unknown) => {
      logger.error("Emotion scan analysis failed", error, "emotion-scan.ui");
      setAnalysis(null);
      const message = error instanceof Error ? error.message : "Analyse impossible pour le moment";
      setErrorMessage(message);
      addError({
        message,
        severity: "high",
        stack: error instanceof Error ? error.stack : undefined,
        context: {
          scope: "emotion-scan-analyze",
          userId: user?.id ?? "anonymous",
        },
      });
    },
  });

  const pa = POS.reduce((s, q) => s + ((resp[q.id] ?? 0) as number), 0); // 0..25
  const na = NEG.reduce((s, q) => s + ((resp[q.id] ?? 0) as number), 0); // 0..25
  const balance = pa - na; // -25..25
  const completion = [...POS, ...NEG].filter(q => !!resp[q.id]).length / 10; // 0..1

  function labelForBalance(b: number) {
    if (b >= 8) return "Équilibre positif";
    if (b >= 3) return "Plutôt positif";
    if (b > -3) return "Neutre";
    if (b > -8) return "Tendu";
    return "Négatif";
    }

  async function onSubmit(e: React.FormEvent) {
    e.preventDefault();
    setSubmitted(true);
    setErrorMessage(null);

    const positiveAnswers = POS.map(question => ({ ...question, value: resp[question.id] as number | undefined }));
    const negativeAnswers = NEG.map(question => ({ ...question, value: resp[question.id] as number | undefined }));

    const summaryText = summarizeLikertAnswers(positiveAnswers, negativeAnswers, balance);

    try {
      const result = await emotionScanMutation.mutateAsync({
        text: summaryText,
        context: "Auto-évaluation I-PANAS-SF",
        previousEmotions: latestScores,
        accessToken: session?.access_token ?? null,
        pa,
        na,
        balance,
      });

      try {
        recordEvent?.({
          module: "emotion-scan",
          startedAt: new Date().toISOString(),
          endedAt: new Date().toISOString(),
          durationSec: 60,
          score: result.emotionalBalance,
          meta: { pa, na, balance, mood: result.dominantEmotion },
        });
      } catch {}
    } catch {
      // handled in mutation onError
    }

    setTimeout(() => document.getElementById("scan-result")?.scrollIntoView({ behavior: "smooth", block: "start" }), 10);
  }

  return (
    <main aria-label="Emotion Scan">
      <PageHeader title="Emotion Scan" subtitle="Auto-évaluation rapide (I-PANAS-SF)" />
      <Card>
        <form onSubmit={onSubmit}>
          <fieldset>
            <legend>Émotions positives</legend>
            {POS.map(q => (
              <LikertRow key={q.id} id={q.id} label={q.label} value={resp[q.id]} onChange={(v)=>setResp(r => ({...r, [q.id]: v}))} />
            ))}
          </fieldset>

          <fieldset style={{ marginTop: 12 }}>
            <legend>Émotions négatives</legend>
            {NEG.map(q => (
              <LikertRow key={q.id} id={q.id} label={q.label} value={resp[q.id]} onChange={(v)=>setResp(r => ({...r, [q.id]: v}))} />
            ))}
          </fieldset>

          <div style={{ display: "grid", gap: 8, marginTop: 12 }}>
            <ProgressBar value={completion * 100} max={100} />
            <Button
              type="submit"
              data-ui="primary-cta"
              aria-disabled={completion < 1 || emotionScanMutation.isPending}
              disabled={completion < 1 || emotionScanMutation.isPending}
            >
              {emotionScanMutation.isPending ? "Analyse en cours…" : "Calculer"}
            </Button>
          </div>
        </form>
      </Card>

      {(submitted || emotionScanMutation.isPending || analysis) && (
        <Card id="scan-result" style={{ marginTop: 12, display: "grid", gap: 12 }}>
          <header>
            <h2>Résultat &amp; analyse IA</h2>
            <p style={{ marginTop: 4, color: "var(--muted-foreground)" }}>
              Synthèse de votre auto-évaluation et recommandations générées par l'IA.
            </p>
          </header>

          <section aria-live="polite" style={{ display: "grid", gap: 4 }}>
            <p><strong>PA</strong> (positif) : {pa} / 25</p>
            <p><strong>NA</strong> (négatif) : {na} / 25</p>
            <p>
              <strong>Balance</strong> (PA - NA) : {balance} — <em>{labelForBalance(balance)}</em>
            </p>
          </section>

          {errorMessage && (
            <p role="alert" style={{ color: "var(--destructive)" }}>
              {errorMessage}
            </p>
          )}

          {emotionScanMutation.isPending && (
            <p style={{ color: "var(--muted-foreground)" }}>Analyse de vos réponses…</p>
          )}

          {analysis && (
            <section aria-label="Analyse IA" style={{ display: "grid", gap: 8 }}>
              <div>
                <p>
                  <strong>Émotion dominante :</strong> {analysis.dominantEmotion}
                </p>
                <p>
                  <strong>Confiance :</strong> {analysis.confidence}% · <strong>Équilibre émotionnel :</strong> {analysis.emotionalBalance}/100
                </p>
              </div>

              <EmotionScoresGrid scores={analysis.emotions} />

              {analysis.insights.length > 0 && (
                <div>
                  <h3>Insights clés</h3>
                  <ul style={{ paddingLeft: 18, marginTop: 4 }}>
                    {analysis.insights.map((insight) => (
                      <li key={insight}>{insight}</li>
                    ))}
                  </ul>
                </div>
              )}

              {analysis.recommendations.length > 0 && (
                <div>
                  <h3>Recommandations personnalisées</h3>
                  <ul style={{ paddingLeft: 18, marginTop: 4 }}>
                    {analysis.recommendations.map((recommendation) => (
                      <li key={recommendation}>{recommendation}</li>
                    ))}
                  </ul>
                </div>
              )}
            </section>
          )}

          <section aria-label="Historique des scans" style={{ display: "grid", gap: 8 }}>
            <h3>Historique des 12 derniers scans</h3>
            {historyLoading && user?.id ? (
              <LoadingState
                variant="section"
                text="Chargement de votre historique émotionnel..."
                skeletonCount={2}
                className="w-full"
              />
            ) : sparklineValues.length ? (
              <div style={{ display: "grid", gap: 8 }}>
                <Sparkline values={sparklineValues} width={320} height={56} />
                {user?.id && historyEntries.length > 0 && (
                  <ul style={{ listStyle: "none", padding: 0, margin: 0, display: "grid", gap: 6 }}>
                    {historyEntries.slice(0, 3).map(entry => (
                      <li key={entry.id} style={{ display: "flex", justifyContent: "space-between", fontSize: "0.9rem" }}>
                        <span>{formatHistoryLabel(entry)}</span>
                        <span>{deriveScore10(entry.normalizedBalance).toFixed(1)}/10</span>
                      </li>
                    ))}
                  </ul>
                )}
              </div>
            ) : (
              <UnifiedEmptyState
                variant="minimal"
                size="sm"
                icon={History}
                title="Aucun historique pour le moment"
                description="Réalisez un scan émotionnel pour suivre votre progression."
                animated={false}
              />
            )}
          </section>
        </Card>
      )}
    </main>
  );
}

function LikertRow({ id, label, value, onChange }: { id: string; label: string; value?: Likert; onChange: (v: Likert)=>void }) {
  const opts: Likert[] = [1,2,3,4,5];
  return (
    <div style={{ display: "grid", gridTemplateColumns: "1fr auto auto auto auto auto", alignItems: "center", gap: 6, paddingBlock: 4 }}>
      <label htmlFor={`${id}-3`} style={{ whiteSpace: "nowrap" }}>{label}</label>
      {opts.map(v => (
        <label key={v} style={{ display: "grid", placeItems: "center" }} aria-label={`${label} ${v}`}>
          <input
            type="radio"
            name={id}
            id={`${id}-${v}`}
            checked={value === v}
            onChange={() => onChange(v)}
          />
          <small>{v}</small>
        </label>
      ))}
    </div>
  );
}

function EmotionScoresGrid({ scores }: { scores: Record<string, number> }) {
  const sorted = React.useMemo(
    () => Object.entries(scores).sort((a, b) => (b[1] ?? 0) - (a[1] ?? 0)).slice(0, 6),
    [scores]
  );

  if (!sorted.length) return null;

  return (
    <div style={{ display: "grid", gap: 4 }}>
      <h3>Scores émotionnels</h3>
      <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fit, minmax(140px, 1fr))", gap: 8 }}>
        {sorted.map(([key, value]) => (
          <div key={key} style={{ border: "1px solid var(--border)", borderRadius: 8, padding: 8 }}>
            <p style={{ fontWeight: 600, textTransform: "capitalize" }}>{key}</p>
            <p style={{ color: "var(--muted-foreground)" }}>{value.toFixed(1)} / 10</p>
          </div>
        ))}
      </div>
    </div>
  );
}

function formatHistoryLabel(entry: EmotionScanHistoryEntry) {
  const date = new Date(entry.createdAt);
  const formatter = new Intl.DateTimeFormat("fr-FR", {
    day: "2-digit",
    month: "short",
    hour: "2-digit",
    minute: "2-digit",
  });

  const dateLabel = formatter.format(date);
  return entry.mood ? `${dateLabel} · ${entry.mood}` : dateLabel;
}
