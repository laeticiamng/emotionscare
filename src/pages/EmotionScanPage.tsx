import React, { useState, useRef, useEffect } from 'react';
import { PageRoot } from '@/components/common/PageRoot';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card';
import { Camera, Mic, Heart, Sparkles, RefreshCw } from 'lucide-react';
import { useMood } from '@/contexts/MoodContext';
import { useActivityLogger } from '@/hooks/useActivityLogger';
import { supabase } from '@/integrations/supabase/client';
import { toast } from '@/hooks/use-toast';

const EmotionScanPage = () => {
  const [scanMode, setScanMode] = useState<'voice' | 'face' | 'mood-cards' | null>(null);
  const [isScanning, setIsScanning] = useState(false);
  const [scanResult, setScanResult] = useState<any>(null);
  const { currentMood, updateMood } = useMood();
  const { logActivity } = useActivityLogger();
  const videoRef = useRef<HTMLVideoElement>(null);
  const mediaStreamRef = useRef<MediaStream | null>(null);

  useEffect(() => {
    logActivity('/scan', 'emotion_scan_page');
  }, [logActivity]);

  const startCameraAccess = async () => {
    try {
      const stream = await navigator.mediaDevices.getUserMedia({ video: true });
      if (videoRef.current) {
        videoRef.current.srcObject = stream;
        mediaStreamRef.current = stream;
      }
      setScanMode('face');
      toast({
        title: "üì∏ Cam√©ra activ√©e",
        description: "Placez votre visage bien en face pour un scan optimal",
      });
    } catch (error) {
      toast({
        title: "Cam√©ra non disponible",
        description: "Utilisons les mood cards √† la place !",
        variant: "destructive",
      });
      setScanMode('mood-cards');
    }
  };

  const startVoiceAccess = async () => {
    try {
      const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
      mediaStreamRef.current = stream;
      setScanMode('voice');
      toast({
        title: "üé§ Micro activ√©",
        description: "Parlez-nous de votre journ√©e pendant 30 secondes",
      });
    } catch (error) {
      toast({
        title: "Micro non disponible",
        description: "Utilisons les mood cards √† la place !",
        variant: "destructive",
      });
      setScanMode('mood-cards');
    }
  };

  const performScan = async () => {
    setIsScanning(true);
    
    try {
      // Call Edge Function for emotion scan
      const { data, error } = await supabase.functions.invoke('emotion-scan', {
        body: {
          mode: scanMode,
          currentMood: currentMood,
        }
      });

      if (error) throw error;

      setScanResult(data);
      
      // Update mood if we got results
      if (data.mood) {
        await updateMood(data.mood.valence, data.mood.arousal);
      }

      toast({
        title: "‚ú® Analyse termin√©e",
        description: data.insight || "Votre √©tat √©motionnel a √©t√© captur√© avec douceur",
      });

    } catch (error) {
      console.error('Scan error:', error);
      toast({
        title: "Erreur temporaire",
        description: "R√©essayons dans un moment",
        variant: "destructive",
      });
    } finally {
      setIsScanning(false);
    }
  };

  const stopMediaStream = () => {
    if (mediaStreamRef.current) {
      mediaStreamRef.current.getTracks().forEach(track => track.stop());
      mediaStreamRef.current = null;
    }
  };

  const resetScan = () => {
    setScanMode(null);
    setScanResult(null);
    stopMediaStream();
  };

  useEffect(() => {
    return () => {
      stopMediaStream();
    };
  }, []);

  const MoodCardsSelector = () => {
    const moodCards = [
      { emoji: 'üòä', label: 'Rayonnant', valence: 0.8, arousal: 0.7 },
      { emoji: 'üòå', label: 'Paisible', valence: 0.6, arousal: 0.3 },
      { emoji: 'üòê', label: 'Neutre', valence: 0.5, arousal: 0.5 },
      { emoji: 'üòî', label: 'Pensif', valence: 0.3, arousal: 0.4 },
      { emoji: 'üò∞', label: 'Stress√©', valence: 0.2, arousal: 0.8 },
      { emoji: 'üò¥', label: '√âpuis√©', valence: 0.4, arousal: 0.2 },
    ];

    return (
      <div className="grid grid-cols-2 gap-4">
        <h3 className="col-span-2 text-lg font-medium text-center mb-4">
          Comment vous sentez-vous maintenant ?
        </h3>
        {moodCards.map((card) => (
          <Button
            key={card.label}
            variant="outline"
            className="h-20 flex flex-col items-center gap-2 hover:bg-primary/10"
            onClick={async () => {
              await updateMood(card.valence, card.arousal);
              setScanResult({
                mood: { valence: card.valence, arousal: card.arousal },
                insight: `Merci d'avoir partag√© que vous vous sentez ${card.label.toLowerCase()}. Nous adaptons l'exp√©rience pour vous.`,
                recommendations: ['Respirations profondes', 'Musique adapt√©e', 'Micro-pause']
              });
            }}
          >
            <span className="text-2xl">{card.emoji}</span>
            <span className="text-sm">{card.label}</span>
          </Button>
        ))}
      </div>
    );
  };

  return (
    <PageRoot className="container mx-auto px-4 py-8">
      <div className="max-w-2xl mx-auto">
        <div className="text-center mb-8">
          <h1 className="text-3xl font-bold mb-4 bg-gradient-to-r from-primary to-primary/60 bg-clip-text text-transparent">
            Scanner de bien-√™tre ‚ú®
          </h1>
          <p className="text-muted-foreground">
            D√©couvrons ensemble votre √©tat du moment pour vous accompagner au mieux
          </p>
        </div>

        {!scanMode && !scanResult && (
          <div className="space-y-4">
            <Card>
              <CardHeader>
                <CardTitle className="flex items-center gap-2">
                  <Camera className="h-5 w-5" />
                  Scan visuel doux
                </CardTitle>
                <CardDescription>
                  Un regard bienveillant sur votre expression du moment
                </CardDescription>
              </CardHeader>
              <CardContent>
                <Button onClick={startCameraAccess} className="w-full">
                  Activer la cam√©ra
                </Button>
              </CardContent>
            </Card>

            <Card>
              <CardHeader>
                <CardTitle className="flex items-center gap-2">
                  <Mic className="h-5 w-5" />
                  √âcoute √©motionnelle
                </CardTitle>
                <CardDescription>
                  Partagez votre voix, nous √©coutons avec attention
                </CardDescription>
              </CardHeader>
              <CardContent>
                <Button onClick={startVoiceAccess} className="w-full">
                  Activer le micro
                </Button>
              </CardContent>
            </Card>

            <Card>
              <CardHeader>
                <CardTitle className="flex items-center gap-2">
                  <Heart className="h-5 w-5" />
                  Cartes d'humeur
                </CardTitle>
                <CardDescription>
                  Une approche simple et intuitive
                </CardDescription>
              </CardHeader>
              <CardContent>
                <Button onClick={() => setScanMode('mood-cards')} variant="outline" className="w-full">
                  Choisir mon humeur
                </Button>
              </CardContent>
            </Card>
          </div>
        )}

        {scanMode === 'face' && !scanResult && (
          <Card>
            <CardHeader>
              <CardTitle>Scan visuel en cours</CardTitle>
              <CardDescription>
                Regardez la cam√©ra avec naturel, le scan prend quelques secondes
              </CardDescription>
            </CardHeader>
            <CardContent className="space-y-4">
              <video
                ref={videoRef}
                autoPlay
                playsInline
                className="w-full h-64 object-cover rounded-lg bg-muted"
              />
              <div className="flex gap-2">
                <Button 
                  onClick={performScan} 
                  disabled={isScanning}
                  className="flex-1"
                >
                  {isScanning ? <RefreshCw className="h-4 w-4 animate-spin mr-2" /> : <Sparkles className="h-4 w-4 mr-2" />}
                  {isScanning ? 'Analyse...' : 'Lancer le scan'}
                </Button>
                <Button variant="outline" onClick={resetScan}>
                  Annuler
                </Button>
              </div>
            </CardContent>
          </Card>
        )}

        {scanMode === 'voice' && !scanResult && (
          <Card>
            <CardHeader>
              <CardTitle>√âcoute active</CardTitle>
              <CardDescription>
                Racontez-nous votre journ√©e, vos ressentis, vos pens√©es...
              </CardDescription>
            </CardHeader>
            <CardContent className="space-y-4">
              <div className="h-32 bg-gradient-to-r from-primary/20 to-secondary/20 rounded-lg flex items-center justify-center">
                <div className="text-center">
                  <Mic className="h-8 w-8 mx-auto mb-2 text-primary animate-pulse" />
                  <p className="text-sm text-muted-foreground">En √©coute...</p>
                </div>
              </div>
              <div className="flex gap-2">
                <Button 
                  onClick={performScan} 
                  disabled={isScanning}
                  className="flex-1"
                >
                  {isScanning ? <RefreshCw className="h-4 w-4 animate-spin mr-2" /> : <Sparkles className="h-4 w-4 mr-2" />}
                  {isScanning ? 'Analyse...' : 'Analyser la voix'}
                </Button>
                <Button variant="outline" onClick={resetScan}>
                  Arr√™ter
                </Button>
              </div>
            </CardContent>
          </Card>
        )}

        {scanMode === 'mood-cards' && !scanResult && (
          <Card>
            <CardHeader>
              <CardTitle>Cartes d'humeur</CardTitle>
              <CardDescription>
                Choisissez celle qui correspond le mieux √† votre √©tat actuel
              </CardDescription>
            </CardHeader>
            <CardContent>
              <MoodCardsSelector />
              <Button variant="outline" onClick={resetScan} className="w-full mt-4">
                Retour aux options
              </Button>
            </CardContent>
          </Card>
        )}

        {scanResult && (
          <Card>
            <CardHeader>
              <CardTitle className="flex items-center gap-2">
                <Sparkles className="h-5 w-5" />
                Analyse termin√©e
              </CardTitle>
            </CardHeader>
            <CardContent className="space-y-4">
              <div className="p-4 bg-gradient-to-r from-primary/10 to-secondary/10 rounded-lg">
                <p className="text-center font-medium">
                  {scanResult.insight}
                </p>
              </div>
              
              {scanResult.recommendations && (
                <div>
                  <h4 className="font-medium mb-2">Suggestions personnalis√©es :</h4>
                  <ul className="space-y-1">
                    {scanResult.recommendations.map((rec: string, index: number) => (
                      <li key={index} className="flex items-center gap-2 text-sm">
                        <span className="w-2 h-2 bg-primary rounded-full"></span>
                        {rec}
                      </li>
                    ))}
                  </ul>
                </div>
              )}

              <div className="flex gap-2">
                <Button onClick={resetScan} variant="outline" className="flex-1">
                  Nouveau scan
                </Button>
                <Button className="flex-1" onClick={() => window.location.href = '/music'}>
                  D√©couvrir la musique adapt√©e
                </Button>
              </div>
            </CardContent>
          </Card>
        )}
      </div>
    </PageRoot>
  );
};

export default EmotionScanPage;