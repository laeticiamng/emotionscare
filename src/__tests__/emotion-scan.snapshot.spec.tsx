import React from "react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
vi.mock("@/COMPONENTS.reg", () => ({
  PageHeader: ({ title }: any) => <div>{title}</div>,
  Button: ({ children, ...props }: any) => <button {...props}>{children}</button>,
  Card: ({ children }: any) => <div>{children}</div>,
  ProgressBar: ({ value }: any) => <div>progress {value}</div>,
  Sparkline: ({ values }: any) => <div>{values.join(',')}</div>
}));
vi.mock("@/contexts/AuthContext", () => ({
  useAuth: () => ({ user: null, isAuthenticated: false }),
}));
vi.mock("@/services/emotionScan.service", () => ({
  invokeEmotionScan: vi.fn().mockResolvedValue({
    emotions: {},
    dominantEmotion: "neutre",
    confidence: 75,
    insights: [],
    recommendations: [],
    emotionalBalance: 50,
  }),
  getEmotionScanHistory: vi.fn().mockResolvedValue([]),
  summarizeLikertAnswers: vi.fn(() => "Résumé"),
  deriveScore10: vi.fn((value: number) => value / 10),
}));
import EmotionScanPage from "@/modules/emotion-scan/EmotionScanPage";

const queryClient = new QueryClient({
  defaultOptions: {
    queries: { retry: false },
    mutations: { retry: false },
  },
});

const wrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);

beforeEach(() => {
  queryClient.clear();
});

describe("EmotionScanPage", () => {
  it("rend la page et le CTA Calculer", () => {
    const { container, getByText } = render(<EmotionScanPage />, { wrapper });
    expect(getByText(/Emotion Scan/i)).toBeTruthy();
    expect(getByText(/Calculer/i)).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
