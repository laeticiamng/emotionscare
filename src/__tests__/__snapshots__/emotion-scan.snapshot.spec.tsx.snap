// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmotionScan Page > rend la page 1`] = `
<div>
  <main
    class="p-4 space-y-4"
  >
    <div>
      EmotionScan
    </div>
    <p>
      Bienvenue dans le module EmotionScan.
    </p>
    <button>
      Commencer
    </button>
  </main>
</div>
`;
